sources/models/src/modeled_types/shared.rs

```diff
@@ -647,6 +647,127 @@ mod test_sysctl_key {
 
 // =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
 
+/// BootConfigKey represents a string that is a valid Kernel boot config keyword; Each key word must
+/// contain only alphabets, numbers, dash (-) or underscore (_).
+/// BootConfigKey stores the original string and makes it accessible through standard traits.
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct BootConfigKeyword {
+    inner: String,
+}
+
+impl TryFrom<&str> for BootConfigKeyword {
+    type Error = error::Error;
+
+    fn try_from(input: &str) -> Result<Self, error::Error> {
+        let valid_keyword = input
+            .chars()
+            .all(|c| (c.is_ascii() && c.is_alphanumeric()) || c == '-' || c == '_')
+            && !input.is_empty();
+        ensure!(
+            valid_keyword,
+            error::InvalidBootconfigKeywordSnafu { input }
+        );
+        Ok(BootConfigKeyword {
+            inner: input.to_string(),
+        })
+    }
+}
+
+string_impls_for!(BootConfigKeyword, "BootConfigKeyword");
+
+#[cfg(test)]
+mod test_bootconfig_keyword {
+    use super::BootConfigKeyword;
+    use std::convert::TryFrom;
+
+    #[test]
+    fn valid_bootconfig_keyword() {
+        for ok in &[
+            "keyword",
+            "key_word",
+            "key-word",
+            "keyword1",
+            "keyword1-",
+            "keyword2_",
+        ] {
+            BootConfigKeyword::try_from(*ok).unwrap();
+        }
+    }
+
+    #[test]
+    fn invalid_bootconfig_keyword() {
+        for err in &[
+            "", "①", ".", "!", "@", "#", "$", "%", "^", "&", "*", "(", ")", "\"", "'", "\\", "|",
+            "~", "`",
+        ] {
+            BootConfigKeyword::try_from(*err).unwrap_err();
+        }
+    }
+}
+
+// =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
+
+/// BootConfigValue represents a string that is a valid Kernel boot config value; Each value only
+/// contains printable characters or spaces except for delimiters such as semicolon, newline, comma,
+/// hash, and closing brace. These delimiters are only usable if the value itself is quoted with
+/// single-quotes or double-quotes.
+/// BootConfigValue stores the original string and makes it accessible through standard traits.
```

Comment:
I am not sure that these quoting rules are applied in the right place, or that the value shape is right.

For command line options like these:
```
slub_debug[=options[,slabs][;[options[,slabs]]...]
usbcore.quirks=0781:5580:bk,0a5c:5834:gij
```

We might input those as:
```
[settings.boot.kernel]
"slub_debug" = [
  "options,slabs",
  "options,slabs",
]
"usbcore.quirks" = [
  "0781:5580:bk",
  "0a5c:5834:gij"
]
```

(The separator between the two is different; `slub_debug` uses semicolons and `usbcore.quirks` uses commas. But that doesn't seem to matter for the purposes of transforming it into valid bootconfig.)

I think what we're saying here is if we had a setting like this:
```
"usbcore.quirks" = [
  ";}",
  "#,"
]
```

Then we'd have to write those bootconfig values as `";}"` and `'#,'` instead of `;}` and `#,`. That feels like a concern we should handle on the serialization side when writing out a value, rather than at this point where we're just concerned with getting tokens that _can_ be written out in a valid way.

On the key side, `usbcore.quirks` is problematic because it's an illegal keyword. I don't think we want to allow all of bootconfig's flexibility, but we do need to support at least one way of inputting data like this.

This seems to match our desired structure pretty well, as long as we allow dots in our keys:
```
foo.bar.baz = value1
foo.bar.qux.quux = value2
```

While this would seem to need a map of keys to any of the following:
* a value
* a list of values
* a nested map of keys
```
foo.bar {
   baz = value1
   qux.quux = value2
}
```

Reply:
For reference, the settings model currently looks like this:
```
// Kernel boot settings
#[model]
struct BootSettings {
    #[serde(...)]
    kernel_parameters: HashMap<BootConfigKeyword, Vec<BootConfigValue>>,
    #[serde(...)]
    init_parameters: HashMap<BootConfigKeyword, Vec<BootConfigValue>>,
}
```

> While this would seem to need a map of keys to any of the following:
>   - a value
>   - a list of values
>   - a nested map of keys

For keys, I'll go ahead replace `BootConfigKeyword` with `BootConfigKey` and allow dotted keys as long as each individual keyword is still valid. I might need to change some things in the serialization module to accommodate, but it should be trivial. I think the deserializing into `BootSettings` might be more complicated since the `Deserializer` would be expecting a nested struct when encountering dotted keys. It's not clear to me at the moment how I best address that.

For values, I'll always assume the value being specified via settings is in its "quoted-form" and then do rudimentary checks on making sure it's ASCII. During serialization to the boot config format, all values will be quoted.

Reply:
I was wondering if I could express the map value as a enum with enum variants mapping to a single boot config value, list of boot config values, or another map. But digging into our datastore serialization and deserialization implementation, we explicitly disallow and error upon tuple variants. So that wouldn't work.
https://github.com/bottlerocket-os/bottlerocket/blob/42bffd64bf9733394a317f961e6ff01852efe184/sources/api/datastore/src/serialization/pairs.rs#L200-L205

Reply:
I think I have a path forward. I forgot I could just implement a custom function to deserialize into `BootSettings` fields like we do for k8s node taints and registry mirrors. Gonna try that out and see.

Reply:
I ended up addressing this by adding a new parameter to the primary deserialization interface that specifies known prefixes for boot config keys. Now keys like `usbcore.quirks` are treated as a single key segment in the datastore.